back-end/src/routes/api.js

```diff
@@ -104,8 +98,203 @@ const uploadFile = (request, response) => {
+      }
+      console.debug("Retrieval successful");
+    }
+    else {
+      response.status(statusCode).send(message);
+    }
+  })().catch(e => {
+    console.debug("Something went wrong");
+  });
+};
+
+/**
+ * Retrieve all photos from a specific tag.
+ * @param {tag: String} request 
+ * @param {[{photo: Object, uploadDate: String, title: String}]} response 
+ */
+const getPhotoByTag = (request, response) => {
+  (async () => {
+    var statusCode = 400;
+    var message = "Empty";
+    const reqTag = request.body.tag;
+    console.debug(reqTag);
+    if (reqTag) {
+      try {
+        var queryParams = {
+          TableName: "Files",
+          IndexName: "tag-file-index",
+          KeyConditionExpression: "#tag = :tag",
+          ExpressionAttributeNames: {
+            "#tag": "tag"
+          },
+          ExpressionAttributeValues: {
+            ":tag": reqTag
+          }
+        };
 
+        const queryData = await dynamodb.query(queryParams).promise();
+        const items = queryData.Items;
+        console.debug("Data Queried");
+        console.debug(items);
+
+        var responseData = [];
+        // Loop through photos to send all object data
+        for (let i = 0; i < items.length; i++) {
+          var getParams = {
+            Bucket: BUCKET_NAME,
+            Key: items[i].file
+          };
+          var photoData = await s3.getObject(getParams).promise();
+            
+          responseObject = {
+            photo: photoData.Body.buffer,
+            title: items[i].title,
+            uploadDate: items[i].uploadDate
+          };
+
+          responseData.push(responseObject);
+        }
+        statusCode = 200;
+        console.debug(responseData);
+        response.status(statusCode).send(responseData);
+      }
+      catch(e) {
+        console.debug("Could not retrieve information", e);
+      }
+    }
+    else {
+      response.status(statusCode).send(message);
+    }
+  })().catch(e => {
+    console.debug("Something went wrong");
+  });
+};
+
+/**
+ * Retrieve all tags and one of their associated photos. 
+ * @param {userID: Integer} request 
+ * @param {[{tag: String, photo: Object}]} response 
+ */
+const getTags = (request, response) => {
+  (async () => {
+    var statusCode = 400;
+    var message = "Nothing to see here folks";
+    // I assume this line is supposed to check if the user is authenticated. Replace once authentication is done.
+    if (true) {
+      try {
+        var scanParams = {
+          TableName: "Files",
+          ProjectionExpression: "#tag, #file",
+          ExpressionAttributeNames: {
+            "#tag": "tag",
+            "#file": "file"
+          }
+        };
+
+        console.debug("Scanning");
+        const scanData = await dynamodb.scan(scanParams).promise();
+        const items = scanData.Items;
+
+        prevTag = "";
+        var responseData = [];
+        for (let i = 0; i < items.length; i++) {
```

Comment:
if the `map` works above please do it here too

Reply:
Getting map to work with async isn't really the issue, it's just that filtering and mapping in one loop this way is a lot easier.